classes/table/accessaudit_table.php

```diff
@@ -0,0 +1,258 @@
+
+        parent::__construct($uniqueid);
+
+        $this->pagesize = $perpage;
+        $this->page = $page;
+        $this->filters = (object)$filters;
+
+        // Define columns in the table.
+        $this->define_table_columns();
+
+        // Define configs.
+        $this->define_table_configs($url);
+
+        // Set download status.
+        $this->is_downloading($download, 'accessaudit_accessaudit_report');
+
+    }
+
+    /**
+     * Setup the headers for the html table.
+     */
+    protected function define_table_columns() {
+        $cols = array(
+            'idnumber' => get_string('idnumber'),
+            'username' => get_string('username'),
+            'email' => get_string('email'),
+            'fullname' => get_string('fullname'),
+            'role' => get_string('role'),
+            'context' => get_string('report_context', 'report_accessaudit'),
+            'contextsystem' => get_string('report_contextsystem', 'report_accessaudit'),
+            'contextuser' => get_string('report_contextuser', 'report_accessaudit'),
+            'contextcoursecat' => get_string('report_contextcoursecat', 'report_accessaudit'),
+            'contextcourse' => get_string('report_contextcourse', 'report_accessaudit'),
+            'contextmodule' => get_string('report_contextmodule', 'report_accessaudit'),
+            'contextblock' => get_string('report_contextblock', 'report_accessaudit')
+        );
+
+        $this->define_columns(array_keys($cols));
+        $this->define_headers(array_values($cols));
+    }
+
+    /**
+     * Define table configs.
+     *
+     * @param \moodle_url $url url of the page where this table would be displayed.
+     */
+    protected function define_table_configs(\moodle_url $url) {
+        $urlparams = (array)$this->filters;
+
+        unset($urlparams['submitbutton']);
+
+        $url->params($urlparams);
+        $this->define_baseurl($url);
+
+        // Set table configs.
+        $this->collapsible(false);
+        $this->sortable(false);
+        $this->pageable(false);
+
+        $this->is_downloadable(true);
+        $this->show_download_buttons_at([TABLE_P_BOTTOM]);
+    }
+
+    /**
+     * Query the reader. Store results in the object for use by build_table.
+     *
+     * @param int $pagesize size of page for paginated displayed table.
+     * @param bool $useinitialsbar do you want to use the initials bar.
+     */
+    public function query_db($pagesize, $useinitialsbar = true) {
+        global $DB;
+
+        $offset = $pagesize * $this->page;
+        $limit = $pagesize;
+
+        list($countsql, $countparams) = $this->get_sql_and_params(true);
+        list($sql, $params) = $this->get_sql_and_params();
+
+        $total = $DB->count_records_sql($countsql, $countparams);
+
+        if ($this->is_downloading()) {
+            $users = $DB->get_records_sql($sql, $params);
```

Comment:
This is very likely to choke at scale, use a recordset like this:

https://github.com/moodle/moodle/blob/master/mod/feedback/classes/responses_table.php#L363-L367